<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обзор данных</a></span><ul class="toc-item"><li><span><a href="#Таблица-books" data-toc-modified-id="Таблица-books-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Таблица <code>books</code></a></span></li><li><span><a href="#Таблица-authors" data-toc-modified-id="Таблица-authors-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Таблица <code>authors</code></a></span></li><li><span><a href="#Таблица-publishers" data-toc-modified-id="Таблица-publishers-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Таблица <code>publishers</code></a></span></li><li><span><a href="#Таблица-ratings" data-toc-modified-id="Таблица-ratings-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Таблица <code>ratings</code></a></span></li><li><span><a href="#Таблица-reviews" data-toc-modified-id="Таблица-reviews-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Таблица <code>reviews</code></a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Анализ-данных" data-toc-modified-id="Анализ-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Анализ данных</a></span><ul class="toc-item"><li><span><a href="#Количество-книг-вышедших-после-1-января-2000-года" data-toc-modified-id="Количество-книг-вышедших-после-1-января-2000-года-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Количество книг вышедших после 1 января 2000 года</a></span></li><li><span><a href="#Количество-обзоров-и-средняя-оценка-для-каждой-книги" data-toc-modified-id="Количество-обзоров-и-средняя-оценка-для-каждой-книги-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Количество обзоров и средняя оценка для каждой книги</a></span></li><li><span><a href="#Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-(менее-50-страниц-считаем-брошюрами)" data-toc-modified-id="Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-(менее-50-страниц-считаем-брошюрами)-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Издательство, которое выпустило наибольшее число книг толще 50 страниц (менее 50 страниц считаем брошюрами)</a></span></li><li><span><a href="#Автора-с-самой-высокой-средней-оценкой-книг,-учитывая-только-книги-с-50-и-более-оценками" data-toc-modified-id="Автора-с-самой-высокой-средней-оценкой-книг,-учитывая-только-книги-с-50-и-более-оценками-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Автора с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками</a></span></li><li><span><a href="#Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок" data-toc-modified-id="Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Среднее количество обзоров от пользователей, которые поставили больше 48 оценок</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

# Анализ базы данных сервиса чтения книг

На волне коронавирусных ограничений компания прибрела крупный сервис для чтения книг по подписке.  
Задачи: 

    Посчитать сколько книг вышло после 1 января 2000 года;
    Для каждой книги посчитать количество обзоров и среднюю оценку;
    Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
    Определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
    Посчитать среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

**Цель проекта:** проанализировать данные для формирования ценностных предложений нового продукта.

**Описание данных**

Таблица books
Содержит данные о книгах:

        book_id — идентификатор книги;
        author_id — идентификатор автора;
        title — название книги;
        num_pages — количество страниц;
        publication_date — дата публикации книги;
        publisher_id — идентификатор издателя.
Таблица authors
Содержит данные об авторах:
 
        author_id — идентификатор автора;
        author — имя автора.
Таблица publishers
Содержит данные об издательствах:

        publisher_id — идентификатор издательства;
        publisher — название издательства;
Таблица ratings
Содержит данные о пользовательских оценках книг:

        rating_id — идентификатор оценки;
        book_id — идентификатор книги;
        username — имя пользователя, оставившего оценку;
        rating — оценка книги.
Таблица reviews
Содержит данные о пользовательских обзорах на книги:
        
        `review_id` — идентификатор обзора;
        `book_id` — идентификатор книги;
        `username` — имя пользователя, написавшего обзор;
        `text` — текст обзора.

## Загрузка данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

Загрузка базы данных

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Обзор данных

Функция `go_query` на вход получает sql запрос и выводит результат в формате таблица pandas.

In [3]:
def go_query(query):
    con=engine.connect()
    return(pd.io.sql.read_sql(sql = query, con = con))

### Таблица `books`

In [4]:
go_query(''' SELECT *
             FROM books
             LIMIT 5
         ''')

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
go_query('''SELECT COUNT(DISTINCT(book_id)) AS books,
                 COUNT(DISTINCT(author_id)) AS authors,
                 COUNT(DISTINCT(publisher_id)) AS publishers,
                 MIN(book_id) as min_book_id,
                 MAX(book_id) as max_book_id,
                 MIN(author_id) as min_author_id,
                 MAX(author_id) as max_author_i,
                 MIN(publisher_id) as min_pub_id,
                 MAX(publisher_id) as max_pub_id,
                 MIN(num_pages) as min_pages,
                 MAX(num_pages) as max_pages,
                 MIN(publication_date) as min_date,
                 MAX(publication_date) as max_date
         FROM books''')

books  authors  publishers  min_book_id  max_book_id  min_author_id  \
0   1000      636         340            1         1000              1   

   max_author_i  min_pub_id  max_pub_id  min_pages  max_pages    min_date  \
0           636           1         340         14       2690  1952-12-01   

     max_date  
0  2020-03-31

В таблице `books` представлена информация о 1000 книг от 636 авторов, изданных в 340 издательствах. МИнимальная длина книги 14 страниц, максимальная 2690. Даты издания варьируются от 1 декабря 1952 года до 31 марта 2020.

### Таблица `authors`

In [6]:
go_query('''  SELECT *
            FROM authors
            LIMIT 5
            ''')

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
go_query('''  SELECT COUNT(DISTINCT(author_id)) AS count_id,
                   COUNT(DISTINCT(author)) AS count_names,
                   MIN(author_id) as min_author_id,
                   MAX(author_id) as max_author_id
            FROM authors
            ''')

count_id  count_names  min_author_id  max_author_id
0       636          636              1            636

`author` количество уникальных идентификаторов и имен авторов 636, то есть у каждого автора есть хтя бы одна книга в таблице `books`

### Таблица `publishers`

In [8]:
go_query('''  SELECT *
            FROM publishers
            LIMIT 5
            ''')

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [9]:
go_query('''  SELECT COUNT(DISTINCT(publisher_id)) AS id_count,
                   COUNT(DISTINCT(publisher)) AS names_pub_counts,
                   MIN(publisher_id) as min_pub_id,
                   MAX(publisher_id) as max_pub_id
            FROM publishers
            ''')

id_count  names_pub_counts  min_pub_id  max_pub_id
0       340               340           1         340

Количество уникальных идентифиаторов издательств совпадает с количеством уникальных имён 340. Идентификаторы от 1 до 340, значит у каждого издательства есть книги в таблице books.

### Таблица `ratings`

In [10]:
go_query('''  SELECT *
            FROM ratings
            LIMIT 5
            ''')

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [11]:
go_query(''' SELECT COUNT(DISTINCT(rating_id)) AS reting_count,
                   COUNT(DISTINCT(book_id)) AS books_count,
                   COUNT(DISTINCT(username)) AS users_count,
                   MIN(book_id) as min_book_id,
                   MAX(book_id) as max_book_id,
                   MIN(rating) as min_rating,
                   MAX(rating) as max_rating
            FROM ratings 
            ''')

reting_count  books_count  users_count  min_book_id  max_book_id  \
0          6456         1000          160            1         1000   

   min_rating  max_rating  
0           1           5

Мы имеем данные о 6456 оценках от 160 пользователей. У всех книг представленных в таблице books есть оценки, минимальный рейтинг 1, максимальный 5.

### Таблица `reviews`

In [12]:
go_query('''  SELECT *
            FROM reviews
            LIMIT 5
            ''')      

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [13]:
go_query('''  SELECT COUNT(DISTINCT(review_id)) AS count_reviews,
                   COUNT(DISTINCT(book_id)) AS count_books,
                   COUNT(DISTINCT(username)) AS count_users,
                   MIN(book_id) as min_book_id,
                   MAX(book_id) as max_book_id,
                   MIN(review_id) as min_review_id,
                   MAX(review_id) as max__review_id
               FROM reviews
            ''')

count_reviews  count_books  count_users  min_book_id  max_book_id  \
0           2793          994          160            1         1000   

   min_review_id  max__review_id  
0              1            2793

Мы имеем 2793 ревью на 994 книги.

### Вывод

Мы ознакомились с представленной для анализа базой, состоящей из 5 таблиц.  
Она содержит информацию о 1000 книг от 636 авторов, изданных в 340 издательствах.  
Длина книг от 14 до 2690 страниц.  
Изданы в период с 1 декабря 1952 года по 31 марта 2020.  
160 пользователей выставили оценки от 1 до 5 баллов всем книгам в базе.   
Ревью есть так же от 160 пользователей у 994 книг.  
Имена авторов и названия издательств уникальны.


## Анализ данных

### Количество книг вышедших после 1 января 2000 года

In [14]:
go_query('''  SELECT COUNT(book_id) AS количество_книг
            FROM books
            WHERE publication_date::date > '2000-01-01'
            ''')

количество_книг
0              819

Количество книг выпущенных после 1 января 2000 года 819.

### Количество обзоров и средняя оценка для каждой книги

In [19]:
go_query('''  SELECT b.title as название,
                   COUNT(DISTINCT(rew.review_id)) AS количество_обзоров,
                   ROUND(AVG(rat.rating),1) AS средняя_оценка
            FROM books AS b LEFT JOIN reviews AS rew ON b.book_id = rew.book_id
            LEFT JOIN ratings AS rat ON b.book_id = rat.book_id
            GROUP BY b.book_id
            ORDER BY 3 DESC
            ''')

название  количество_обзоров  \
0    Crucial Conversations: Tools for Talking When ...                   2   
1    The Walking Dead  Book One (The Walking Dead #...                   2   
2                        A Dirty Job (Grim Reaper  #1)                   4   
3                                            Geek Love                   2   
4                                  Homage to Catalonia                   2   
..                                                 ...                 ...   
995                             The Kitchen God's Wife                   3   
996                  His Excellency: George Washington                   2   
997                                      Drowning Ruth                   3   
998                                              Junky                   2   
999                               Harvesting the Heart                   2   

     средняя_оценка  
0               5.0  
1               5.0  
2               5.0  
3               5.0  
4               5.0  
..              ...  
995             2.3  
996             2.0  
997             2.0  
998             2.0  
999             1.5  

[1000 rows x 3 columns]

### Издательство, которое выпустило наибольшее число книг толще 50 страниц (менее 50 страниц считаем брошюрами)

In [16]:
go_query('''SELECT publisher as издатель, 
                   COUNT(book_id) as количество_книг
            FROM books as b INNER JOIN publishers as p ON p.publisher_id = b.publisher_id
            WHERE num_pages > 50
            GROUP BY publisher
            ORDER BY count(book_id) DESC
            LIMIT 1
            ''')

издатель  количество_книг
0  Penguin Books               42

Наиболее продуктивным в выпуске книг толще 50 страниц является Penguin Books.

### Автора с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками

In [17]:
go_query('''  WITH bb AS (SELECT b.book_id,
                               AVG(rat.rating)
                        FROM ratings AS rat LEFT JOIN books as b ON b.book_id = rat.book_id
                        GROUP BY b.book_id
                        HAVING count(rat.rating) >= 50)
            SELECT a.author_id AS идентификатор_автора,
                    a.author AS автор,
                    ROUND(AVG(bb.avg),1) AS средний_рейтинг
            FROM bb LEFT JOIN books as b ON bb.book_id = b.book_id
            LEFT JOIN authors as a ON a.author_id = b.author_id
            GROUP BY 1
            ORDER BY 3 DESC
            LIMIT 1
            ''')

идентификатор_автора                       автор  средний_рейтинг
0                   236  J.K. Rowling/Mary GrandPré              4.3

### Среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [18]:
go_query('''  WITH users AS (SELECT rat.username
                           FROM ratings AS rat
                           GROUP BY rat.username
                           HAVING count(rat.rating_id) >48),
             count_rew AS (SELECT r.username AS user,
                                  COUNT(r.review_id)
                           FROM users LEFT JOIN reviews AS r ON users.username = r.username
             GROUP BY 1)
            SELECT AVG(count) AS среднее_количество_обзоров
            FROM count_rew
         ''')

среднее_количество_обзоров
0                        24.0

В среднем 24 обзора сделали пользователи, поставившие более 48 оценок.

## Вывод

На вход была получена база данных сервиса для чтения книг по подписке. Мы ознакомились с данными проверили корректность значений.  

В ходе анализа выяснили следующее:
    
       1. после 01 января 2020 года было выпущено 819 книг
       
       2. была выведена таблица с количеством ревью и средней оценкой по всем книгам. Наибольшее количество,7 обзороз, получила книга Сумерки часть 1 ("Twilight (Twilight #1)"), при этом её средняя оценка невысока 3.7 баллов. Ни одного обзора и самая низкая средняя оценка в 3 балла у книги "Естественный путь к рисованию" ("The Natural Way to Draw")
       
       3.издательство Penguin Books выпустило наибольшее количество книг толще 50 страниц, а именно 42 книги. Продуктивность издательства неудивительна. Издательство основано в 1935 году в Лондоне, главной миссией являлось сделать бумажные книги доступными. Однако с 1993 года начало выпускать и аудиокниги. В 2013 году объединилось со своим главным конкурентом из США, создав первую в мире глобальную издательскую компанию. 
       
       4.среди книг с 50ю и более оценками, автором с самой высокой средней оценкой являются Джоан Роулинг/Мэри Грандпре J.K. Rowling/Mary GrandPré. Джоан Роулинг автор, а Мэри Грандпре иллюстратор серии книг о Гарри Поттере. Такое количество оценок и достаточно высокий средний рейтинг 4.3 неудивительны, у книг множество фанатов по всеми миру.
       
       5. в среднем 24 обзора сделали пользователи, поставившие более 48 оценок.

       
    
    